### PR TITLE
feat!: Switch to aztec_backend that uses upstream BB & UltraPlonk

### LINKED ISSUE #991: [Nargo] Upgrade to the UltraPlonk Version of Aztec Backend
### Problem

The current `aztec_backend` and `aztec_wasm_backend` Nargo uses are based on the TurboPlonk version of [Aztec Backend](https://github.com/noir-lang/aztec_backend), which lacks efficient implementations of useful primitives (e.g. Keccak256 in 18k constraints, ECDSA verification in 36k constraints) that the UltraPlonk version offers.

### Proposed solution

Switch to the UltraPlonk version of Aztec Backend.

### Alternatives considered

_No response_

### Additional context

At the moment of submitting this issue, this is known to be blocked by certain upgrade works on both [Aztec Backend](https://github.com/noir-lang/aztec_backend) and [Barretenberg](https://github.com/AztecProtocol/barretenberg/) needed for integration with Noir, including but not limited to:

```[tasklist]
### Tasks
- [ ] https://github.com/AztecProtocol/barretenberg/pull/299
- [ ] https://github.com/noir-lang/aztec_backend/pull/84
- [ ] https://github.com/AztecProtocol/barretenberg/pull/338
- [ ] https://github.com/noir-lang/barretenberg-sys/pull/12
- [ ] https://github.com/AztecProtocol/barretenberg/pull/350
- [ ] https://github.com/noir-lang/noir/pull/1114
```

The issue is created more for tracking purposes.

### LINKED ISSUE #1105: Nix Nargo Builds

### LINKED ISSUE #1115: Upgrade acvm to v0.9.0
### Problem

Version 0.9 of acvm has been released here: https://github.com/noir-lang/acvm/pull/182

### Proposed solution

Upgrade to v0.9 

This is blocked by backends being updated to v0.9. See https://github.com/noir-lang/aztec_backend/issues/104

### Alternatives considered

_No response_

### Additional context

_No response_

### Submission Checklist

- [X] Once I hit submit, I will assign this issue to the Project Board with the appropriate tags.

### PR DESCRIPTION
**Pull request by phated**
# Related issue(s)

Resolves #991
Resolves #1001
Resolves #1105 
Resolves #1115 

# Description

This updates the `aztec_backend` dependencies to the latest commit, which has started depending on upstream barretenberg and switches to UltraPlonk.

Since BB has a lot of dependencies to be built, @kobyhallx and I have added a Nix flake to the repository to manage this dependency.

I've updated the workflows to use Nix to run clippy and cargo tests.

This is currently a draft because more work needs to be done to update values in the tests, and it seems that there is a CRS bug in BB because we are getting a panic when trying to run the prove_and_verify tests. Will need to work with @vezenovm and @kevaundray to sort that out.

## Test additions / changes

# Checklist

- [x] I have tested the changes locally.
- [x] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` with default settings.
- [x] I have [linked](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue) this PR to the issue(s) that it resolves.
- [x] I have reviewed the changes on GitHub, line by line.
- [x] I have ensured all changes are covered in the description.

## Documentation needs
- [x] This PR requires documentation updates when merged.

# Additional context

Docs updates:
1. We compile to ultra plonk now
2. Building from source drastically changed (follow up work is to make `nix profile install ...` to install nargo globally)

### CODE REVIEW

File: crates/nargo_cli/tests/test_data/8_integration/src/main.nr
Comment (phated) [inline crates/nargo_cli/tests/test_data/8_integration/src/main.nr:253]: @vezenovm what's happening here?

File: crates/nargo_cli/src/cli/codegen_verifier_cmd.rs
Comment (phated) [inline crates/nargo_cli/src/cli/codegen_verifier_cmd.rs:17]: Is this changing the codegen verifier command? cc @TomAFrench

File: crates/nargo/tests/test_data/testing/src/main.nr
Comment (phated) [inline crates/nargo/tests/test_data/testing/src/main.nr:1]: @vezenovm are we supposed to move these to nargo_cli?

File: crates/nargo_cli/src/lib.rs
Comment (phated) [inline crates/nargo_cli/src/lib.rs:2]: @TomAFrench you were talking about this at the off-site. I found https://github.com/rust-lang/rust/issues/72686 which says that it's so buggy they dropped it from default commands and called it "fixed" but someone commented and said it was still broken.

I removed it because it causes the nix build to fail.

File: crates/nargo_cli/tests/prove_and_verify.rs
Comment (phated) [inline crates/nargo_cli/tests/prove_and_verify.rs:45]: This was very misleading so I made it more clear

File: Cargo.toml
Comment (phated) [inline Cargo.toml:55]: Need to remove once bb is working upstream

File: flake.nix
Comment (phated) [inline flake.nix:23]: need to use a commit on master once bb is working upstream

**Comment by phated** [comment]
Blocked by https://github.com/AztecProtocol/barretenberg/pull/338 and https://github.com/noir-lang/barretenberg-sys/pull/12 (which is also waiting on a resolution to 338)

File: crates/nargo_cli/tests/test_data/8_integration/src/main.nr
Comment (vezenovm) [reply crates/nargo_cli/tests/test_data/8_integration/src/main.nr:253]: This is residual debugging from when we were iterating on bberg with UP. Will submit a PR to clean this stuff up

File: crates/nargo/tests/test_data/testing/src/main.nr
Comment (vezenovm) [reply crates/nargo/tests/test_data/testing/src/main.nr:1]: We can probably remove this test. Or move this to `nargo_cli` to have a pure logic constraint test. We also test logic constraints in `8_integration` just a lot more is going on in that test.

Comment (vezenovm) [reply crates/nargo/tests/test_data/testing/src/main.nr]: I moved it to an `xor` test under `nargo_cli` lmk what you think

File: crates/nargo_cli/tests/test_data/8_integration/src/main.nr
Comment (vezenovm) [reply crates/nargo_cli/tests/test_data/8_integration/src/main.nr]: I reverted the `8_integration` test

**Comment by Globallager** [comment]
Are https://github.com/noir-lang/noir/issues/1069 and https://github.com/noir-lang/noir/issues/1074 similar to the CRS bug encountered?

File: crates/nargo_cli/src/cli/codegen_verifier_cmd.rs
Comment (TomAFrench) [reply crates/nargo_cli/src/cli/codegen_verifier_cmd.rs:17]: In what way? It should be a non-breaking change as we''l fall back to the original behaviour if no circuit name is passed.

**Comment by phated** [comment]
@Globallager no, those seem to be due to a bad/buggy library that aztec_backend is using to download the CRS. We need to replace it but it hasn't been a high priority.

File: crates/wasm/src/circuit.rs
Comment (phated) [inline crates/wasm/src/circuit.rs:26]: @TomAFrench @kobyhallx Are we fine to remove these deprecated functions with the acvm 0.9.0 upgrade?

Comment (TomAFrench) [reply crates/wasm/src/circuit.rs:26]: Yes, we're not using this anymore and they got removed as they've been deprecated since 0.2.1: https://github.com/noir-lang/acvm/pull/177

File: crates/nargo_cli/src/cli/codegen_verifier_cmd.rs
Comment (TomAFrench) [inline crates/nargo_cli/src/cli/codegen_verifier_cmd.rs:27]: I'd like to add a "context" struct at some point which essentially wraps the `PackageManifest` along with the package's root directory. These sorts of functions would then live as methods on there.

File: .gitignore
Comment (phated) [inline .gitignore:5]: @kevaundray why was this ignored? We want to allow it to recommend plugins and default some settings in the workspace.

File: crates/nargo_cli/src/cli/codegen_verifier_cmd.rs
Comment (phated) [inline crates/nargo_cli/src/cli/codegen_verifier_cmd.rs:1]: @TomAFrench I had to resolve some conflicts from your changes, can you make sure they look good?

File: crates/nargo_cli/tests/prove_and_verify.rs
Comment (phated) [inline crates/nargo_cli/tests/prove_and_verify.rs:1]: @TomAFrench conflicts in this file too, please double check

File: .gitignore
Comment (kevaundray) [reply .gitignore:5]: I think this was a part of my default gitignore or a gitignore template

File: crates/nargo_cli/src/cli/codegen_verifier_cmd.rs
Comment (TomAFrench) [inline crates/nargo_cli/src/cli/codegen_verifier_cmd.rs:40]: Can you switch this to use the `codegen_verifier` function from `nargo` lib?

File: crates/nargo_cli/src/cli/codegen_verifier_cmd.rs
Comment (TomAFrench) [reply crates/nargo_cli/src/cli/codegen_verifier_cmd.rs:1]: Made a new comment with the suggested change.

File: crates/nargo_cli/tests/prove_and_verify.rs
Comment (TomAFrench) [reply crates/nargo_cli/tests/prove_and_verify.rs:1]: Looks good to me.

File: crates/nargo_cli/src/cli/codegen_verifier_cmd.rs
Comment (phated) [reply crates/nargo_cli/src/cli/codegen_verifier_cmd.rs]: Done

**Comment by kevaundray** [comment]
First Mac OS runner has passed CI checks :)

**Comment by phated** [comment]
I'll need to adjust the required checks again once we have approvals.

File: crates/nargo_cli/src/cli/codegen_verifier_cmd.rs
Comment (kevaundray) [inline crates/nargo_cli/src/cli/codegen_verifier_cmd.rs:24]: Can you create an issue for this instead of the TODO?

File: crates/nargo_cli/tests/prove_and_verify.rs
Comment (kevaundray) [inline crates/nargo_cli/tests/prove_and_verify.rs:53]: This seems to revert some of the changes by @TomAFrench in #1180 -- Both refactors make the code more readable, so I'm okay with keeping either

File: cspell.json
Comment (kevaundray) [inline cspell.json:59]: "mybranch"  being here indicates that a variable should be named "my_branch" or "myBranch" somewhere

File: cspell.json
Comment (kevaundray) [inline cspell.json:58]: I'm guessing there is somewhere that says `TODO(phated)` -- if so, can we create an issue and remove from cspell

File: flake.nix
Comment (kevaundray) [inline flake.nix:87]: How does this work for larger circuits whom may need more points?

File: crates/nargo_cli/tests/prove_and_verify.rs
Comment (kevaundray) [inline crates/nargo_cli/tests/prove_and_verify.rs:8]:
```suggestion
    use std::path::{Path, PathBuf};
```

File: README.md
Comment (kevaundray) [inline README.md:100]: Not a blocker for this PR: should we have a central place where we put nix and direnv instructions? It seems like we will have duplicated information between all of our repositories which will use nix.

**Review by kevaundray — COMMENTED** [review]
Generally looks good to me -- Left some comments -- I skipped over the nix config parts

File: crates/nargo_cli/tests/prove_and_verify.rs
Comment (TomAFrench) [inline crates/nargo_cli/tests/prove_and_verify.rs:60]:
```suggestion
        // Load config.tml file from test_data directory
```
nit

File: crates/nargo_cli/tests/prove_and_verify.rs
Comment (TomAFrench) [reply crates/nargo_cli/tests/prove_and_verify.rs:53]: This is fine and addresses https://github.com/noir-lang/noir/pull/1114/files/6a79a0e94260a78d5ccd5c25b24c657b4ebac7f2#r1162150191

This doesn't really affect my refactoring as that focused on the for-loop below.

File: cspell.json
Comment (kevaundray) [reply cspell.json:58]: I don't actually see `phated` anywhere, so going to remove and see if spellcheck screams at me

Comment (kevaundray) [inline cspell.json:58]:
```suggestion
```

File: README.md
Comment (kevaundray) [inline README.md:131]:
```suggestion
nix flake lock --override-input barretenberg github:username/barretenberg/my_branch
```

File: cspell.json
Comment (kevaundray) [reply cspell.json:58]: Found the usage in readme :)

File: cspell.json
Comment (kevaundray) [inline cspell.json:58]:
```suggestion
```

**Review by TomAFrench — DISMISSED** [review]
Everything looks good to me. Tested everything seems to work.

File: flake.nix
Comment (vezenovm) [reply flake.nix:87]: I think our backend only currently downloads transcript00 at the moment anyway. Perhaps we should add a new issue to handle larger circuits that need more points and thus a larger transcript

File: flake.nix
Comment (vezenovm) [inline flake.nix:168]: Do we have an issue for this TODO?

File: flake.nix
Comment (vezenovm) [inline flake.nix:150]: And an issue for this one here?

**Review by vezenovm — DISMISSED** [review]
Some minor comments

File: flake.nix
Comment (TomAFrench) [reply flake.nix:87]: This is something that @kobyhallx is currently thinking about btw.

Comment (TomAFrench) [reply flake.nix:87]: Actually... can we not remove this @phated? This seems like it's only relevant in the flake for `aztec_backend`.

File: flake.nix
Comment (phated) [reply flake.nix:168]: #1197

File: flake.nix
Comment (TomAFrench) [reply flake.nix:87]: Do we not inherit from when it's set in `aztec_backend` though? That said, we also need to specify `barretenberg` even though that's only used in `aztec_backend` too.

In any case it's not a blocker and I can experiment later.

File: flake.nix
Comment (phated) [reply flake.nix:150]: #1198

File: README.md
Comment (phated) [reply README.md:100]: #1199

File: crates/nargo_cli/src/cli/codegen_verifier_cmd.rs
Comment (phated) [reply crates/nargo_cli/src/cli/codegen_verifier_cmd.rs:24]: #1201

**Review by kevaundray — APPROVED** [review]

File: flake.nix
Comment (phated) [reply flake.nix:87]: Unfortunately (or maybe fortunately ⁉️ ), we don't inherit anything from flakes in any rust projects because they are consumed as crates, not flakes.

**Pull request merged**
